_posts/2020-08-03-the-missing-combine-input-type.md

```diff
@@ -0,0 +1,368 @@
+
+    let subject = PassthroughSubject<Parameters, Never>()
+}
+```
+
+There's a fair bit going on in this little snippet so let's unpack it quickly...
+
+The reason I first thought PropertyWrappers might be a good solution is because they allow us to expose things differently. 
+We can use the publicly visible members (`wrappedValue` and `projectedValue`) to expose the parameters for the input and a way to send them. 
+We can them use the restricted members (the `_underscore` accessor) to give the enclosing type exclusive access to the `Subject`.
+
+We don't really want the `wrappedValue` here, it's just a means to expose a generic parameter. 
+The `projectedValue` exposes a `SendProxy` which is just a way for use to provide _only_ the `send` function to code outside the view model.
+
+Clear as mud? Let's take a look at how we might use this in our view model and hopefully it'll make sense:
+
+```swift
+class AuthViewModel {
+    // MARK: - Outputs
+    let signInResult: AnyPublisher<Result<Token, Error>, Never>
+
+    // MARK: - Inputs
+    @Input var signIn: (username: String, password: String)
+
+    // MARK: - Lifecycle
+    init() {
+        signInResult = _signIn // underscore accessor let's us use the `Subject` exclusively
+            .subject
+            .flatMap { input in
+                return API
+                    .authenticate(input.username, input.password)
+                    .materialize()
+            }
+            .eraseToAnyPublisher()
+    }
+}
+```
+
+We can now sue the following syntax to send values:
+
+```swift
+viewModel.$signIn.send((username: "...", password: "..."))
+```
+
+The PropertyWrapper itself is a little crazy, but the view model and the call site is a little cleaner I think. 
+Unfortunately there is actually a pretty big downside to this approach, PropertyWrappers can't be enforced in protocols.
+
+Consider this:
+
+```swift
+protocol AuthViewModelType {
+    // MARK: - Outputs
+    var signInResult: AnyPublisher<Result<Token, Error>, Never> { get }
+
+    // MARK: - Inputs
+    @Input var signIn: (username: String, password: String) { get } // Property 'signIn' declared inside a protocol cannot have a wrapper
+}
+```
+
+Since we can't enforce `@Input` it means code that only knows about the `AuthViewModelType` wouldn't have visibility to call the `$signIn.send`  function.
+
+This is a bit of a show stopper... I don't usually use protocols for view models but for other Combine friendly dependencies I might create I want to be able to enforce strict input/outputs.
+
+Take two... let's try using a concrete type:
+
+```swift
+struct AnyConsumer<Output> {
+    private let subject = PassthroughSubject<Output, Never>()
+
+    func send(_ value: Output) {
+        subject.send(value)
+    }
+}
+```
+
+I'm calling this `AnyConsumer` because it feels like a nice parallel with `AnyPublisher`.  This type allows us to expose _just_ the `send` function making this type _input only_ just like we want... 
```

Comment:
```suggestion
I'm calling this `AnyConsumer` because it feels like a nice parallel with `AnyPublisher`.  This type allows us to expose _just_ the `send` function making this type _input only_ just like we want…
```